### PR TITLE
feat(headless): support relative dates in date facet

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-819

What remains:
- Serialization for search parameters (currently only supports absolute values)
- New unit tests (adapted the ones we had)
- Atomic Timeframe Facet support (done, PR not opened)

Lots of modified file because of the following renames:
- `DateRangeRequest` -> `DateRangeApiRequest`
- `DateFacetValue` -> `DateFacetApiValue`
- The orignal interfaces are now extended to support relative ranges

### CODE REVIEW

File: packages/headless/doc-parser/use-cases/search.ts [hunk 1]
```diff
@@ -366,6 +366,9 @@ const actionLoaders: ActionLoaderConfiguration[] = [
   {
     initializer: 'loadDateFacetSetActions',
   },
+  {
+    initializer: 'laodRelativeDateSetActions',
```
Comment (olamothe) [inline packages/headless/doc-parser/use-cases/search.ts:370]:
```suggestion
    initializer: 'loadRelativeDateSetActions',
```

File: packages/headless/src/features/relative-date-set/relative-date-actions-loader.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  ): PayloadAction<RegisterRelativeDateActionCreatorPayload>;
+}
+
+/**
+ * Loads the `dateFacetSet` reducer and returns possible action creators.
+ *
+ * @param engine - The headless engine.
+ * @returns An object holding the action creators.
+ */
+export function laodRelativeDateSetActions(
```
Comment (olamothe) [inline packages/headless/src/features/relative-date-set/relative-date-actions-loader.ts:32]:
```suggestion
export function loadRelativeDateSetActions(
```

**Review by olamothe — APPROVED** [review]

File: packages/headless/src/features/relative-date-set/relative-date-actions-loader.ts (hunk 2 above)
Comment (ThibodeauJF) [reply packages/headless/src/features/relative-date-set/relative-date-actions-loader.ts:32]: At least the typo is consistent! 😆

File: packages/headless/src/features/relative-date-set/relative-date-set-state.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import {RelativeDate} from './relative-date';
+
+export type RelativeDateSetState = Record<string, Record<string, RelativeDate>>;
```
Comment (btaillon) [inline packages/headless/src/features/relative-date-set/relative-date-set-state.ts:3]: Why do we need a new reducer and state? Would it be possible to register date ranges from relative dates instead?

**Review by btaillon — APPROVED** [review]

File: packages/headless/src/features/relative-date-set/relative-date-set-state.ts (hunk 3 above)
Comment (ThibodeauJF) [reply packages/headless/src/features/relative-date-set/relative-date-set-state.ts:3]: What you're proposing is similar to the original solution: https://github.com/coveo/ui-kit/pull/973

File: packages/headless/src/features/relative-date-set/relative-date-set-slice.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  getRelativeDateSetInitialState(),
+  (builder) => {
+    builder.addCase(registerRelativeDate, (state, {payload}) => {
+      const {id, relativeDate, absoluteDate} = payload;
+      if (!state[id]) {
+        state[id] = {};
+        return;
+      }
+
+      state[id][absoluteDate] = relativeDate;
```
Comment (samisayegh) [inline packages/headless/src/features/relative-date-set/relative-date-set-slice.ts:15]: Could it be possible for two different relative dates to collide/have the same absoluteDate, causing an unexpected resolution? For example, 1 week vs. 7 days, where the ranges are:

```
7days ago - now
2 weeks ago - 1 week ago
```

Depending on the order, `1 week ago` might be resolved as `7 days ago` in the controller. Someone rendering the ranges using the controller state would unexpectedly print: `2 weeks ago - 7 days ago`.

If this is possible, it might be safer to store the full range, so that it is possible to compare the start and end, to ensure we are resolving the relative date the developer originally configured.

File: packages/headless/src/controllers/facets/range-facet/date-facet/headless-date-facet-options.ts [hunk 5]
```diff
@@ -126,7 +160,7 @@ export const dateFacetOptionsSchema = new Schema<Required<DateFacetOptions>>({
 
 export function validateDateFacetOptions(
   engine: SearchEngine<ConfigurationSection & SearchSection & DateFacetSection>,
-  options: DateFacetOptions
+  options: RegisterDateFacetActionCreatorPayload
```
Comment (samisayegh) [inline packages/headless/src/controllers/facets/range-facet/date-facet/headless-date-facet-options.ts:163]: Similar to above, shouldn't the validation be on the full facet options? or should the function name be adjusted?
If the latter, could we reuse the validation logic inside the `registerDateFacet` action?

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 import {FacetValueState} from '../../../../features/facets/facet-api/value';
+import {formatDateForSearchApi} from '../../../../api/date-format';
+import {RelativeDate} from '../../../../features/relative-date-set/relative-date';
+import {DateRangeRequest} from '../../../../features/facets/range-facets/date-facet-set/interfaces/request';
 
 dayjs.extend(utc);
-dayjs.extend(customParseFormat);
 
-export type DateRangeInput = string | number | Date;
+export type DateRangeInput = string | number | Date | RelativeDate;
```
Comment (ThibodeauJF) [inline packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts:10]: Modified this module in order to be able to pass a RelativeDate as well as another type of range.
Previously in my other branch we had a new buildRelativeRange module instead. But now mixed ranges are supported (start is absolute, end is relative)

File: packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts [hunk 7]
```diff
@@ -127,27 +152,56 @@ export function validateManualDateRanges(
...
-        `The start value is greater than the end value for the date range ${value.start} to ${value.end}`
+        `The start value is greater than the end value for the date range ${serializeDateRangeValue(
+          value.start
+        )} to ${serializeDateRangeValue(value.end)}`
       );
     }
   });
 }
 
+function formatRelativeDateValues(
```
Comment (ThibodeauJF) [inline packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts:165]: I used to use formatDate in both slices: DateFacetSet & RelativeDateSet, but this sometimes caused issues where the absolute date was slightly different (around 1 ms).
Instead I enhance the payload in order to have the map of absolute value + relative value that both slices can reuse. Single source of truth

File: packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts [hunk 8]
```diff
@@ -187,7 +241,7 @@ export interface UpdateDateFacetValuesActionCreatorPayload {
   /**
    * The date facet values.
    */
-  values: DateFacetValue[];
+  values: DateRangeRequest[];
```
Comment (ThibodeauJF) [inline packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts:244]: Type was wrong here

File: packages/headless/src/features/facets/range-facets/date-facet-set/interfaces/request.ts [hunk 9]
```diff
@@ -27,7 +33,19 @@ export interface DateRangeRequest {
   state: FacetValueState;
 }
 
+export interface DateRangeMappedRequest extends DateRangeRequest {
```
Comment (ThibodeauJF) [inline packages/headless/src/features/facets/range-facets/date-facet-set/interfaces/request.ts:36]: Used to pass the values from the action to the slices.
Will also be used to return the search parameters

File: packages/headless/src/features/search-parameters/search-parameter-serializer.ts [hunk 10]
```diff
@@ -159,6 +209,7 @@ function buildNumericRanges(ranges: string[]) {
 }
 
 function buildDateRanges(ranges: string[]) {
+  // TODO: parse relative ranges from string
```
Comment (ThibodeauJF) [inline packages/headless/src/features/search-parameters/search-parameter-serializer.ts:212]: Missing piece of the puzzle

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts [hunk 11]
```diff
@@ -47,8 +47,12 @@ export interface DateRangeOptions {
...
 }
 
-export function isSearchApiDate(date: string) {
-  return formatForSearchApi(dayjs(date)) === date;
+function isRelativeDate(rawDate: DateRangeInput): rawDate is RelativeDate {
+  return (
+    typeof rawDate !== 'number' &&
+    typeof rawDate !== 'string' &&
+    !(rawDate instanceof Date)
+  );
```
Comment (samisayegh) [inline packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts:55]: Instead of determining a relative date by omission, it would be safer to check if it has a `period`, `units` etc.
Otherwise, if a new type is added to `DateRangeInput`, this function will need to be adjusted,

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts [hunk 12]
```diff
@@ -71,7 +79,7 @@ export function buildDateRange(config: DateRangeOptions): DateRangeRequest {
   };
 }
 
-function buildDate(rawDate: DateRangeInput, options: DateRangeOptions) {
+function buildDate(rawDate: string | number | Date, options: DateRangeOptions) {
```
Comment (samisayegh) [inline packages/headless/src/controllers/facets/range-facet/date-facet/date-range.ts:82]: If we store this type as `AbsoluteDate`, then `DateRangeInput` could become `AbsoluteDate | RelativeDate`

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts [hunk 13]
```diff
@@ -232,13 +232,26 @@ function getNumericFacets(state: Partial<SearchParametersState>) {
...
+      const selectedRanges = getSelectedRanges(request.currentValues).map(
+        (facetValue) => {
+          return {
+            ...facetValue,
+            start:
+              relativeDates.find(({value}) => facetValue.start === value) ||
+              facetValue.start,
+            end:
+              relativeDates.find(({value}) => facetValue.end === value) ||
+              facetValue.end,
```
Comment (samisayegh) [inline packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:251]: Wouldn't the start and end stored in relativeDates be identical to those in the dateFacetSet? If so, then why not simplify to always read `start` and `end` from the dateFacetSet?

The relativeDateSet would be used to fill the relativeStart and relativeEnd,

File: packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts [hunk 14]
```diff
@@ -161,7 +215,7 @@ export interface ToggleSelectDateFacetValueActionCreatorPayload {
   /**
    * The target date facet value.
    */
-  selection: DateFacetValue;
+  selection: DateFacetApiValue;
```
Comment (samisayegh) [inline packages/headless/src/features/facets/range-facets/date-facet-set/date-facet-actions.ts:218]: Why this change? From the perspective of a developer using Headless, they would only see `DateFacetValue` when using the controller. I think `DateFacetApiValue` is an internal concept that should not be exposed?

File: packages/headless/src/features/search-parameters/search-parameter-serializer.ts [hunk 15]
```diff
@@ -158,13 +216,60 @@ function buildNumericRanges(ranges: string[]) {
...
+  } else {
+    const matches = date.match(relativeDateFormatRegexp)!;
+    relativeDate = {
+      period: matches[1] as RelativeDatePeriod,
+      amount: parseInt(matches[2]),
+      unit: matches[3] as RelativeDateUnit,
+    };
+  }
+
+  return {...relativeDate, value: formatDate(relativeDate)};
```
Comment (ThibodeauJF) [inline packages/headless/src/features/search-parameters/search-parameter-serializer.ts:250]: I'm having an issue here with the serialization. Since every time that we deserialize we have new absolute values, and the date-facet-set-slice does not contain the relative date, we cannot compare the values correctly. A workaround would be to provide the deserializer with the maps so he returns the previous absolute value instead of a new one.

File: packages/headless/src/features/facets/range-facets/date-facet-set/interfaces/request.ts (hunk 9 above)
Comment (samisayegh) [reply packages/headless/src/features/facets/range-facets/date-facet-set/interfaces/request.ts:36]: Would be simpler to extend `DateRangeApiRequest` instead of `DateRangeRequest`, and add keys to that interface? e.g.

```
export interface DateRangeMappedRequest extends DateRangeApiRequest {
    relativeStart?: RelativeDate;
    relativeEnd?: RelativeDate;
}
```

File: packages/headless/src/features/search-parameters/search-parameter-serializer.ts (hunk 15 above)
Comment (ThibodeauJF) [reply packages/headless/src/features/search-parameters/search-parameter-serializer.ts:250]: Or I would need to add the relative date maps to the search params 🤔

File: packages/headless/src/features/relative-date-set/relative-date-set-slice.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  if (!state[facetId]) {
+    state[facetId] = [];
+  }
+
+  values.forEach((value) => {
+    addIfRelative(value.start, state[facetId]);
+    addIfRelative(value.end, state[facetId]);
+  });
```
Comment (samisayegh) [inline packages/headless/src/features/relative-date-set/relative-date-set-slice.ts:49]: Would it be possible to avoid the loops, and store the `DateRangeMappedRequest[]` as-is?
